src/circulating_supply/circulating_supply_provider.rs

```diff
@@ -0,0 +1,150 @@
+use bigdecimal::{BigDecimal, ToPrimitive};
+use diesel::PgConnection;
+
+use crate::circulating_supply::lockup::{
+    get_account_state, get_code_version, is_bug_inside, TRANSFERS_ENABLED,
+};
+use crate::circulating_supply::user_balance::get_user_balance;
+use crate::db_adapters::accounts::collect_lockups_for_block;
+use crate::db_adapters::blocks::closest_block_for;
+use crate::db_adapters::circulating_supply::{
+    add_circulating_supply, get_precomputed_circulating_supply,
+};
+use crate::models;
+use crate::models::circulating_supply::CirculatingSupply;
+use crate::models::Block;
+use actix::Addr;
+use actix_web::rt::time;
+use std::ops::Div;
+use std::str::FromStr;
+use std::time::SystemTime;
+use tracing::info;
+
+const DAY: std::time::Duration = std::time::Duration::from_secs(86400);
+
+pub(crate) async fn compute_circulating_supply(view_client: Addr<near_client::ViewClientActor>) {
+    let mut current_day = TRANSFERS_ENABLED;
+    while current_day < get_now_timestamp() {
+        compute_circulating_supply_for(&view_client, current_day).await;
+        current_day += DAY.as_nanos() as u64;
+    }
+
+    // We ping the number twice a day to be sure that we computed it, and to deliver it earlier.
+    // If the value is already in the DB, we will just return almost immediately
+    let mut interval = time::interval(DAY.div(2));
+    loop {
+        compute_circulating_supply_for(&view_client, get_now_timestamp()).await;
+        interval.tick().await;
+    }
+}
+
+fn get_now_timestamp() -> u64 {
+    SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_nanos() as u64
+}
+
+async fn compute_circulating_supply_for(
+    view_client: &Addr<near_client::ViewClientActor>,
+    request_timestamp: u64,
+) {
+    let pool = models::establish_connection();
```

Comment:
Let's establish a database connections pool once for the whole indexer. In the worst case, establish the pool for all the background jobs once

Reply:
It's better now, but I still clone it sometimes. I tried
```rust
let pool: &'static Database<PgConnection> = &models::establish_connection();
```
It solves the problem with passing the value to any threads, but it gives an error
```
error[E0716]: temporary value dropped while borrowed
   --> src/main.rs:220:50
    |
220 |     let pool: &'static Database<PgConnection> = &models::establish_connection();
    |               -------------------------------    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary which is freed while still in use
    |               |
    |               type annotation requires that borrow lasts for `'static`
...
298 | }
    | - temporary value is freed at the end of this statement
```
Tried to fight with it: no chance for me